aiida/backends/djsite/db/migrations/0024_dblog_update.py

```diff
@@ -6,39 +6,216 @@
+
+            # If delete_on_close is False, we are running for the user and add additional message of file location
+            if not delete_on_close:
+                click.echo('Exported legacy workflow logs to {}'.format(filename))
+
+            # Now delete the records
+            DbLog.objects.exclude(objname__startswith=node_prefix).exclude(
+                objname__startswith=leg_workflow_prefix).delete()
+
+
+def copy_objpk_data_to_dbnode_column(apps, _):
+    """
+    Copy the (clean) objpk data to the dbnode column which is a foreign key to the dbnode
+    """
+    from django.db import transaction
+
+    DbLog = apps.get_model('db', 'DbLog')
+    with transaction.atomic():
+        for log in DbLog.objects.all():
+            log.dbnode_id = log.objpk
+            log.save(update_fields=['dbnode_id'])
+
+
+def clean_dblog_metadata(apps, _):
+    """
+    Remove objpk and objname from DbLog table.
+    """
+    from django.db import transaction
+    import json
+
+    DbLog = apps.get_model('db', 'DbLog')
+    with transaction.atomic():
+        for log in DbLog.objects.all():
+            met = json.loads(log.metadata)
+            if 'objpk' in met:
+                del met['objpk']
+            if 'objname' in met:
+                del met['objname']
+            log.metadata = json.dumps(met)
+            log.save(update_fields=['metadata'])
+
 
 class Migration(migrations.Migration):
     """
-    This migration updated the DbLog schema and adds UUID for correct export of the DbLog entries.
+    This migration updates the DbLog schema and adds UUID for correct export of the DbLog entries.
+    More specifically, it adds UUIDS, it exports to files the not needed log entries (that correspond
+    to legacy workflows and unknown entities), it creates a foreign key to the dbnode table, it
+    transfers there the objpk data to the new dbnode column (just altering the objpk column and making
+    it a foreign key when containing data, raised problems) and in the end objpk and objname columns
+    are removed.
     """
 
     dependencies = [
         ('db', '0023_calc_job_option_attribute_keys'),
     ]
 
     operations = [
-        migrations.RemoveField(
+        # We export export the logs of the old workflows to a JSON file
+        migrations.RunPython(export_and_clean_workflow_logs, reverse_code=migrations.RunPython.noop),
+
+        # We create a new field called dbnode which is a foreign key to the dbnode table
+        migrations.AddField(
             model_name='dblog',
-            name='objpk',
+            name='dbnode',
+            field=models.ForeignKey(
+                on_delete=models.deletion.CASCADE, related_name='dblogs', to='db.DbNode', blank=True, null=True),
         ),
-        migrations.AddField(
+        # We move all the data from the objpk to the node field
+        migrations.RunPython(copy_objpk_data_to_dbnode_column, reverse_code=migrations.RunPython.noop),
```

Comment:
Then we should also recreate the `objname` which technically we could get from the associated node's type string, but do we really want to go through this effort? It will delay this PR even more and plus, we already cannot fully reverse since we have lost the workflow log entries

Reply:
Mmm I see.
No, definitely I wouldn't delay this PR longer. But can we at least make sure that if a person migrates back and forth, the ending state does not imply losing data? Now it would, I think, as going back would lose the association to the node. In this spirit, it's ok if going back does not recreate legacy workflows and the correct objname string.
Otherwise one should 'block/prevent' downgrading (but I don't know if this e.g. creates problems in the test infrastructure that moves a lot between versions)

Reply:
In general, as it is now, the test don't complain.
I did not create any data migration when downgrading (I just create the correct schema - that's why all the tests pass) because we lose data anyway.

I am not sure if it makes a lot of sense to copy the dbnode_id to objpk without having correctly populated the objname. We have also other migrations that the downgrade doesn't work.

In any case, I would avoid having a downgrade migration that doesn't half of the job in migarting data. I would prefer to warn the users that there is no "way back"